### PR TITLE
http: add removeIf() header map function

### PR DESCRIPTION
**Pull request by mattklein123**
To be used by out of tree extensions.

Risk Level: Low
Testing: New tests and existing
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @jessicayuen @LisaLudique

File: source/common/http/header_map_impl.cc [hunk 1]
```diff
@@ -525,6 +507,18 @@ size_t HeaderMapImpl::removePrefix(const LowerCaseString& prefix) {
...
 
+size_t HeaderMapImpl::remove(const LowerCaseString& key) {
+  return HeaderMapImpl::removeIf([&key](const HeaderEntry& entry) -> bool {
+    return key.get() == entry.key().getStringView();
+  });
+}
+
+size_t HeaderMapImpl::removePrefix(const LowerCaseString& prefix) {
+  return HeaderMapImpl::removeIf([&prefix](const HeaderEntry& entry) -> bool {
+    return absl::StartsWith(entry.key().getStringView(), prefix.get());
```
Comment (jmarantz) [inline source/common/http/header_map_impl.cc:518]: just a quick note that we can make both of these functions fast if we use std::map as a lazy associative structure (as opposed to a hash-map).

File: source/common/http/header_map_impl.cc [hunk 2]
```diff
@@ -483,28 +483,10 @@ void HeaderMapImpl::clear() {
   cached_byte_size_ = 0;
 }
 
-size_t HeaderMapImpl::remove(const LowerCaseString& key) {
+size_t HeaderMapImpl::removeIf(const std::function<bool(const HeaderEntry&)>& predicate) {
```
Comment (jmarantz) [inline source/common/http/header_map_impl.cc:486]: consider making `using HeaderEntryPredicate` here or similar.

**Comment by mattklein123** [comment]
@jmarantz updated

File: source/common/http/header_map_impl.cc [hunk 3]
```diff
@@ -483,7 +483,7 @@ void HeaderMapImpl::clear() {
   cached_byte_size_ = 0;
 }
 
-size_t HeaderMapImpl::removeIf(const std::function<bool(const HeaderEntry&)>& predicate) {
+size_t HeaderMapImpl::removeIf(const HeaderMap::HeaderMatchPredicate& predicate) {
```
Comment (jmarantz) [inline source/common/http/header_map_impl.cc:486]: FWIW I think you can drop the HeaderMap qual here as HeaderMapImpl is a prefix.

Comment (mattklein123) [reply source/common/http/header_map_impl.cc:486]: It's not actually (it's a mixin).

Comment (jmarantz) [reply source/common/http/header_map_impl.cc:486]: got it; makes sense.

**Comment by rgs1** [comment]
Hmm I am seeing appends in the `:Authority` header (e.g.: `:authority = foo,bar` -- which did not happened before) after this change... could it be related?

**Comment by rgs1** [comment]
Just confirmed -- reverting this change makes the appends in `:authority` go away. I'll dig in.

**Comment by rgs1** [comment]
Ok -- per @fishcakez this is probably related to the fact that we historically used `request_headers_to_add` to manipulate the `host` header, when we should have used `host_rewrite`.

So that's probably the issue, I'll confirm in a bit. We might still want to do something about this, since it might bite others.

**Comment by mattklein123** [comment]
Hmm, I wouldn't think there would be a behavior change here but `remove()` did change slightly to not check for it being an O(1) header and to do a full scan. I can fix this if we think this is going to be an issue. @rgs1 can you provide a more complete description of the flow causing the issue?

**Comment by rgs1** [comment]
Instead of using `host_rewrite` we had:

```
             "request_headers_to_add": [
              {
                "append": false,
                "header": {
                  "key": "host",
                  "value": "api.foo.com"
                }
              },
```

After this change, `append: false` isn't honored anymore and we get `:authority = foo,bar`.

I think it's more of our mistake for manipulating the `host` header directly instead of using `host_rewrite`, than
a regression since this is technically an incorrect way of doing a host rewrite.

If we don't fix it we could at least clarify the docs or disallow referencing the `host` header
in `requests_headers_to_add`? Happy to send a PR for either one.

**Comment by mattklein123** [comment]
OK let me look at this a bit and get back to you.

**Comment by mattklein123** [comment]
OK I see the issue, basically the old path was doing the host -> authority translation and the new path is not.

@rgs1 I agree to be on the safe side I will go ahead and fix this. Can I fix this tomorrow or do you need this reverted urgently?

**Comment by rgs1** [comment]
> OK I see the issue, basically the old path was doing the host -> authority translation and the new path is not.
> 
> @rgs1 I agree to be on the safe side I will go ahead and fix this. Can I fix this tomorrow or do you need this reverted urgently?

No rush on our end, we got unblocked by moving over to using `host_rewrite`. Thanks for the quick return!

**Pull request merged**